### PR TITLE
feat(go): Add Prometheus monitoring example

### PR DESCRIPTION
**Pull request by kahlys**
Add a prometheus example in Go

### CODE REVIEW

File: src/go/monitor/main.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+import (
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+func main() {
+	http.Handle("/metrics", promhttp.Handler())
+	http.ListenAndServe(":2112", nil)
```
Comment (satreix) [inline src/go/monitor/main.go:11]: Can we get this address from a configurable place, like a flag?

File: src/go/monitor/main.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package main
+
+import (
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+func main() {
+	http.Handle("/metrics", promhttp.Handler())
```
Comment (satreix) [inline src/go/monitor/main.go:10]: Can we extract this server into a separate struct and add a test?

Comment (satreix) [inline src/go/monitor/main.go:10]: Do we want to add a simple visit counter middleware for the entire app, to show how to expose a metric?

**Review by satreix — COMMENTED** [review]
This all looks good. Thanks for the contribution.

Please:

- add a little blurb in the PR description about what this does and where you wan to take this next?
- have a look at [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/); I think this would fit nicely as: "feat(go): Add prometheus example"

File: src/go/monitor/main.go (hunk 2 above)
Comment (kahlys) [reply src/go/monitor/main.go]: I will add a metric to count the total of visit.
To keep in simple, for example purpose, I propose to keep a single `main.go`file.

Comment (kahlys) [reply src/go/monitor/main.go]: I added a simple test using `httptest`

File: src/go/monitor/main.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	flag.Parse()
+
+	service := NewService()
+
+	serv := http.Server{
+		Addr:    *addr,
+		Handler: service.Handler(),
+	}
+
+	log.Printf("Listenning on %s", *addr)
```
Comment (satreix) [inline src/go/monitor/main.go:67]:
```suggestion
	log.Printf("Listening on http://%s", *addr)
```

File: src/go/monitor/main.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+// serviceMetrics holds metrics for a service
+type serviceMetrics struct {
+	VisitTotal prometheus.Counter
+}
+
+// newServiceMetrics registers metrics
+func newServiceMetrics() serviceMetrics {
+	return serviceMetrics{
+		VisitTotal: promauto.NewCounter(prometheus.CounterOpts{
+			Name: "myapp_visit_total",
```
Comment (satreix) [inline src/go/monitor/main.go:22]: Can we extract `myapp` into a top level constant and add a param to this `newServiceMetrics` top be used as prefix to all the metrics?

File: src/go/monitor/main.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+func NewService() Service {
+	return Service{
+		metrics: newServiceMetrics(),
+	}
+}
+
+// Handler return a HTTP handler for the service.
+func (s Service) Handler() http.Handler {
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.Handler())
```
Comment (satreix) [inline src/go/monitor/main.go:43]: Can we add a simple root page to show that the service works? Nothing fancy a "OK" or a link to the metrics page would be better then a 404 😃

File: src/go/monitor/main.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	return s.middleware(mux)
+}
+
+// middleware to log and record metrics
+func (s Service) middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		log.Println(r.Method, r.RequestURI)
+		s.metrics.VisitTotal.Inc()
```
Comment (satreix) [inline src/go/monitor/main.go:51]: If this middleware logs and update the metrics on all routes; does it need to be attached to the full `Service`, or can we wrap the server with this middleware at the `ListenAndServe` level instead?

File: src/go/monitor/main_test.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+func TestVisitTotal(t *testing.T) {
+	s := NewService()
+
+	request, err := http.NewRequest(http.MethodGet, "/metrics", nil)
+	if err != nil {
+		t.Fatal("creating request:", err)
+	}
+
+	expected := 1
```
Comment (satreix) [inline src/go/monitor/main_test.go:22]: Can this be const?

**Review by satreix — COMMENTED** [review]
This very clean! A few touch ups and we are good to go.

File: src/go/monitor/main.go (hunk 6 above)
Comment (kahlys) [reply src/go/monitor/main.go:51]: It was just a basic middleware as an example. We can remove the logging part. For the monitoring part, my preferences are:
- If we want to monitor the service, keep the method attach to the Service struct
- If we only want to monitor the http handler/router/etc, use the middleware at the ListenAndServe level.

File: src/go/monitor/main.go (hunk 3 above)
Comment (kahlys) [reply src/go/monitor/main.go:67]: To stay coherent, a modification will be to be made here too https://github.com/satreix/everest/blob/main/src/go/echo/server/server.go#L37

**Review by satreix — APPROVED** [review]
Thanks for the contribution!

**Pull request merged**